### ISSUE TITLE
Bug: web animation cancel() followed by animate() behave oddly when requestAnimationFrame's involved

### ISSUE THREAD

**Issue by chenglou** [facebook/react#31634]
See repro: [tinyurl.com/4u85ehnh](https://t.co/3eMtHJfEQ5)

Video of the glitch: https://x.com/_chenglou/status/1860978002643063054

Tldr on button click I cancel the existing animation and in `useLayoutEffect` I start a new one. The animation goes from translate 100px to 200px. However, clicking on the button will show a split second glitched unstyled view of the animated text (position at 0px) before jumping to 100px and animating to 200px. Commenting out the `requestAnimationFrame` wrapper inside `handleClick` solves it.

Happens across Chrome, Safari and Firefox.

Notes:
- The rAF happens on the same frame as everything else, so there shouldn't be any visible DOM repaint
- Even if the rAF is executed later, it'd means `cancel()` is executed _later_, which should hide the unstyled visual, not reveal it
- Although, the fact is that rAF can be _scheduled_ earlier, if this is relevant: https://x.com/nomsternom/status/1853687984266055983 (original post unrelated to this issue)
- You can check the difference in console.logs. The microtask log result & order is different. The logs themselves (including getComputedStyle which forces reflows) don't disturb this repro
- FWIW, I was wondering whether the first frame of `animate()` is skipped when `cancel()` is done in the rAF, and it seems it's preserved. Another debugging video (first 6 times glitching, 7th time not) below. The first frame's position is 100px, which is where the red square is. For alignment. Here:

https://github.com/user-attachments/assets/c7a9a21f-57e7-4123-bac5-8a07a02741f4

I'd also like to be able to get a clean, React-less repro but not sure how complicated that'd be.

**Comment by pixelwarrior0ff** [comment]
1. Contrary to popular belief, Lorem Ipsum is not simply random text. 'It has roots in a piece of classical Latin literature from 45 BC, making it over 2000 years at Hampden-Sydney College in Virginia, l
2. ooked up one of the more obscure Latin words, consectetur, from a Lorem Ipsum passage, and going 'hbvh
4. through the cites of the word in classical literature, discovered the undoubtable source. Lorem Ipsum comes from sections 1.10.32 and 1.10.33 of "de Finibus Bonorum et Malorum" (The Extremes of Good and Evil) bs
5. 
6. y Cicero, written in 45 BC. This book is a treatise on the theory of ethics, very popular during the Renaissance. The first line of Lorem Ipsum, "Lorem ipsum dolor sit amet..", comes from a line in section 1.10.32.
7. 
8. The standard chunk of Lorem Ipsum used since the 1500s is reproduced below for those interested. Sections 1.10.32 and 1.10.33 from "de Finibus Bonorum et Malorum" by Cicero are also reproduced in their exact original form, accompanied by English versions from the 1914 translation by H. Rackham.

**Comment by pixelwarrior0ff** [comment]
1. Contrary to popular belief, Lorem Ipsum is not simply random text. 'It has roots in a piece of classical Latin literature from 45 BC, making it over 2000 years at Hampden-Sydney College in Virginia, looked up one of the more obscure Latin words, consectetur, from a Lorem Ipsum passage, and going 'hbvh
                                                          2. 
                                                          3. through the cites of the word in classical literature, discovered the undoubtable source. Lorem Ipsum comes from sections 1.10.32 and 1.10.33 of "de Finibus Bonorum et Malorum" (The Extremes of Good and Evil) bs
                        4. 
                        5. y Cicero, written in 45 BC. This book is a treatise on the theory of ethics, very popular during the Renaissance. The first line of Lorem Ipsum, "Lorem ipsum dolor sit amet..", comes from a line in section 1.10.32.
                        6. 
                        7. The standard chunk of Lorem Ipsum used since the 1500s is reproduced below for those interested. Sections 1.10.32 and 1.10.33 from "de Finibus Bonorum et Malorum" by Cicero are also reproduced in their exact original form, accompanied by English versions from the 1914 translation by H. Rackham.

**Comment by sebmarkbage** [comment]
I think you need to wrap forceUpdate in flushSync. Because in rAF the updates have “default” priority which means get batched into the next rAF and not the currently executing one.

It’s a special case for updates within rAF that if you need them to execute within the rAF itself they need to be flush synced.

It should be the default if we had any way to detect that we’re in a rAF like we do for events like click, mouse down/up etc.

**Comment by chenglou** [comment]
Oh I see, without `rAF`, `cancel()` ended up happening this frame, `forceUpdate` happens next, then `useLayoutEffect` also happens next.

Clearer repro for folks looking for this: https://tinyurl.com/4yz8nes8

Btw is there a place in the docs to document some of these scheduling stuff? I can do it
